sklearn/metrics/_classification.py

```diff
@@ -143,6 +147,32 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
         return sample_score.sum()
 
 
+def _nan_average(scores, weights):
+    """
+    Wrapper to combine np.average and np.nanmean, so np.nan values are ignored
+    from the average and weights can be passed. Note that when possible,
+    we degenerate to the prime methods.
+    """
+
+    if len(scores) == 0:
+        return np.nan
+
+    mask = np.isnan(scores)
+    if mask.all():
+        return np.nan
+
+    if weights is None:
+        return np.nanmean(scores)
```

Comment:
A bit link with the comment of @thomasjpfan, I am wondering if we should precompute `mask.any()` and use the basic mean when no `nan` are present. It should avoid some overhead:

```python
is_any_nam = mask.any()
if weights is None:
    return np.nanmean(scores) if is_any_nan else np.mean(scores)
elif not is_any_nan:
    return np.average(scores, weights=weights)
```

Reply:
this does not work well because when scores sums 0, np.average throws a `ZeroDivisionError`

Reply:
Oh yes, right.